### PR TITLE
Fix arange when step is negative

### PR DESCRIPTION
**Pull request by naoyam**
The device version of ceilDiv assumes positive inputs, so when step is negative, it gives an incorrect result. For example, I see FusionStandAloneArange results in a write error with compute-sanitizer when start = 0, stop = -1, step = -1.5 and dtype = kLong. Here's sanitizer output:

```
========= Invalid __global__ write of size 8 bytes
=========     at 0x310 in CudaCodeGen::kernel5(long long, long long, long long, CudaCodeGen::Tensor<long long, (int)1>)
=========     by thread (1,0,0) in block (0,0,0)
=========     Address 0x7f2fbc600808 is out of bounds
=========     and is 1 bytes after the nearest allocation at 0x7f2fbc600800 of size 8 bytes
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame: [0x21740c]
```

We could fix the device ceilDiv, but I'm feeling we don't want to introduce the conditional code if possible.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -471,7 +471,9 @@ TensorView* arange(Val* start, Val* end, Val* step, DataType dtype) {
     end = castOp(DataType::Double, end);
     step = castOp(DataType::Double, step);
   }
-  auto size = castOp(DataType::Int, ceilDiv(sub(end, start), step));
+  // Make sure no negative value is passed to ceilDiv as the device
+  // implementation of ceilDiv assumes positive inputs
+  auto size = castOp(DataType::Int, ceilDiv(abs(sub(end, start)), abs(step)));
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.cpp:476]: Actually, I'm not really sure what the right behavior should be when the signs of `end - start` and `step` are different, which I think is invalid. Do we need to do error checking? Doing so in device code could be very costly, so we probably would want to do on the host side before launching kernels. 

In any case, unless we want to check it on device, the above workaround should be fine.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp:476]: I think we do need to check it on the host side, but I am not sure how. Would it be a good idea to have both a `BinaryOpType::CeilDiv` and a `BinaryOpType::CeilDivMaybeNegative`, and we use `CeilDivMaybeNegative` for arange, and use `CeilDiv` for others?

File: torch/csrc/jit/codegen/cuda/dynamic_type.h [hunk 2]
```diff
@@ -186,6 +186,14 @@ class TORCH_CUDA_CU_API IntOrDouble {
   explicit operator int64_t() const;
   explicit operator size_t() const;
   explicit operator int() const;
+
+  IntOrDouble abs() const {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/dynamic_type.h]: Should we make this a function inside `namespace IntOrDouble_functions` in this file to be consistent with other functions?

**Review by zasdfgbnm — APPROVED** [review]
LGTM, thanks for fixing!

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp:476]: I don't know if we would need additional ceilDiv expr types. I think it should be possible to determine if it's guaranteed to be safe or not by looking at Expr nodes, and we could just do host-side checks of those ceilDiv exprs that are not determined to be safe.

That said, I don't think it's important at this point. At least for now, I think it'd be just reasonable if nvFuser works if a given program is correct but could result in undefined behavior if an unsound program is given. We would definitely want to leave a note about the lack of the check.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp:476]: But, if it should be validated anyway, I think we would want to extend `KernelIrScanner` to mark potentially unsafe ceilDiv exprs. As I mentioned above, I think we should be able to just analyze the input vals of those exprs and see if they are composed of just non-negative values such as domain extents.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp:476]: A note is fine for now. We can revisit in the future if needed.

**Pull request merged**